### PR TITLE
Copy aufs hardlinks to top layer

### PR DESCRIPTION
**Pull request by tonistiigi**
Copy aufs hardlinks to top layer  and show warning if old method is used.

---

Docker used to support AUFS hardlink references in layer tars. These references live in `.wh..wh.plnk` directory and are hardlinks to files connecting them to the inodes in the aufs mount. This makes them temporary and unusable for keeping the hardlink relationship with other files outside the layer. Instead they may cause collisions with other files when reused and require weird hacks on other graphdrivers because although those files are not used, their contents need to be accessed during extraction because they are linked to other files.

The situation became worse in `v1.8` with the introduction of tar-split. `plnk` files will be kept in the `tar-split` JSON but because they are temporary their contents can't be read back from the mount when one tries to reassemble the tar, resulting a failed push. Before these cases would just result a different tar during push. (#15643)

Since `v1.4` `plnk` files are not not included in the tar creation. (#8813)

This fix brings hardlink support between different layers for AUFS by always copying changed hardlinks to a top layer before creating the tar. This should be the preferred method for storing hardlinks in layers. It doesn't fix the same issue for other graphdrivers that would require manual scanning for this files.

Also we now show a warning in daemon logs when extracting anything that contains `plnk`. I would actually like to make it a hard failure but we probably need to wait for a next version for this.

Notice that just removing the `plnk` extraction hacks without failing wouldn't be a solution because then people would end up with wrong files if they unpack an old layer.

cc @vbatts @jlhawn

### CODE REVIEW

**Comment by vbatts** [comment]
SGTM (i have not done a code review yet)

**Comment by LK4D4** [comment]
@tonistiigi Failure on janky seems related.

**Comment by tonistiigi** [comment]
Had to remove the warning on non-aufs apply so this is hardly ideal now. Seems our `chrootarchive` package does not let log messages through. I tried to pipe stderr to main process but some tests expect the output to be captured in the error messages.

**Comment by vbatts** [comment]
Is that because of the reexec?

**Comment by tonistiigi** [comment]
@vbatts Its because the extraction runs in another chrooted process (via reexec) and stdio for the subcommand is used internally for error handling, not forwarded to the main process stdio.

File: daemon/graphdriver/aufs/aufs.go [hunk 1]
```diff
@@ -350,6 +350,10 @@ func (a *Driver) Put(id string) error {
 // Diff produces an archive of the changes between the specified
 // layer and its parent layer which may be "".
 func (a *Driver) Diff(id, parent string) (archive.Archive, error) {
+	// Make sure the hardlinks are copied up before taking the diff
+	if err := exec.Command("auplink", path.Join(a.rootPath(), "mnt", id), "flush").Run(); err != nil {
```
Comment (vbatts) [inline daemon/graphdriver/aufs/aufs.go]: do you think this might be better in `func a.unmount` since this flush is needed before every unmount?

Comment (tonistiigi) [inline daemon/graphdriver/aufs/aufs.go]: I think its already called on Unmount, https://github.com/docker/docker/blob/master/daemon/graphdriver/aufs/mount.go#L14

Comment (vbatts) [inline daemon/graphdriver/aufs/aufs.go]: Oh I see. We're covered there as well.

File: daemon/graphdriver/aufs/aufs.go [hunk 2]
```diff
@@ -350,6 +350,10 @@ func (a *Driver) Put(id string) error {
...
+		logrus.Debugf("Couldn't run auplink before unmount: %s", err)
```
Comment (tiborvass) [inline daemon/graphdriver/aufs/aufs.go]: shouldnt this be a warning?

File: daemon/graphdriver/aufs/aufs.go [hunk 3]
```diff
@@ -360,10 +364,20 @@ func (a *Driver) Diff(id, parent string) (archive.Archive, error) {
 }
 
 func (a *Driver) applyDiff(id string, diff archive.Reader) error {
-	return chrootarchive.UntarUncompressed(diff, path.Join(a.rootPath(), "diff", id), &archive.TarOptions{
+	dir := path.Join(a.rootPath(), "diff", id)
```
Comment (tiborvass) [inline daemon/graphdriver/aufs/aufs.go]: i know it was already `pth` but i wonder why it's not `filepath`

**Comment by tiborvass** [comment]
@tonistiigi do you think we can have a test for this one?

File: daemon/graphdriver/aufs/aufs.go (hunk 2 above)
Comment (tonistiigi) [inline daemon/graphdriver/aufs/aufs.go]: Its `Errorf` in `Unmount`. What do you prefer?

Comment (tiborvass) [inline daemon/graphdriver/aufs/aufs.go]: Oh so are we going to see the error twice? Let's make it an error here too i guess if we have to keep both.

**Comment by tonistiigi** [comment]
@tiborvass Unit-test added

File: daemon/graphdriver/aufs/aufs.go (hunk 3 above)
Comment (tonistiigi) [inline daemon/graphdriver/aufs/aufs.go]: There are 18 mentions of `path.*` in this file. That change needs to be in another PR.

Comment (tiborvass) [inline daemon/graphdriver/aufs/aufs.go]: Ok makes sense

**Comment by icecrime** [comment]
:see_no_evil: 

```
+ go test -test.timeout=60m github.com/docker/docker/daemon/graphdriver/aufs
time="2015-10-13T01:01:45Z" level=error msg="Couldn't run auplink before diff: exit status 22" 
time="2015-10-13T01:01:45Z" level=error msg="Couldn't run auplink before diff: exit status 22" 
time="2015-10-13T01:01:45Z" level=error msg="Couldn't run auplink before diff: exit status 22" 
time="2015-10-13T01:01:45Z" level=error msg="Couldn't run auplink before diff: signal: segmentation fault (core dumped)" 
time="2015-10-13T01:01:46Z" level=error msg="Couldn't run auplink before unmount: signal: segmentation fault (core dumped)" 
--- FAIL: TestHardlinks (0.62s)
	aufs_test.go:719: Target files are not linked
FAIL
```

**Comment by tiborvass** [comment]
I have the same error locally

**Comment by tonistiigi** [comment]
I took a different approach and now the test works on janky but it probably makes sense to move the milestone to `1.10` because I don't know what caused the auplink segfaults and if noplnk option has any side effects.

**Comment by icecrime** [comment]
Cool let's do that then, thanks @tonistiigi!

**Comment by vbatts** [comment]
so is this ready for merge otherwise? since 1.9 is frozen and cherry picking at this point, we can land this in master, right?

**Comment by thaJeztah** [comment]
ping @tiborvass @tonistiigi @icecrime is this ready to go?

@vbatts is that a LGTM from your side?

**Comment by LK4D4** [comment]
LGTM

**Comment by vbatts** [comment]
LGTM
